gengokit/gentesthelper/helper.go

```diff
@@ -0,0 +1,67 @@
+package gentesthelper
+
+import (
+	"bytes"
+	"fmt"
+	"go/ast"
+	"go/parser"
+	"go/printer"
+	"go/token"
+	"reflect"
+	"runtime"
+	"strings"
+
+	"github.com/pmezard/go-difflib/difflib"
+)
+
+// FuncSourceCode returns a string representing the source code of the function
+// provided to it.
+func FuncSourceCode(val interface{}) (string, error) {
+	ptr := reflect.ValueOf(val).Pointer()
+	fpath, _ := runtime.FuncForPC(ptr).FileLine(ptr)
+
+	funcName := runtime.FuncForPC(ptr).Name()
+	parts := strings.Split(funcName, ".")
+	funcName = parts[len(parts)-1]
+
+	// Parse the go file into the ast
+	fset := token.NewFileSet()
+	fileAst, err := parser.ParseFile(fset, fpath, nil, parser.ParseComments)
+	if err != nil {
+		return "", fmt.Errorf("ERROR: go parser couldn't parse file '%v'\n", fpath)
+	}
+
+	// Search ast for function declaration with name of function passed
+	var fAst *ast.FuncDecl
+	for _, decs := range fileAst.Decls {
+		switch decs.(type) {
+		case *ast.FuncDecl:
+			f := decs.(*ast.FuncDecl)
+			if f.Name.String() == funcName {
+				fAst = f
+			}
+		}
+	}
+	code := bytes.NewBuffer(nil)
+	err = printer.Fprint(code, fset, fAst)
+
+	if err != nil {
+		return "", fmt.Errorf("couldn't print code for func '%v': %v\n", funcName, err)
+	}
+
+	return code.String(), nil
```

Comment:
Is it actually important that this be returned as a `string`, or would `[]byte` actually be okay? In many ways, `[]byte` is more flexible than `string`. Early on in my Go tenure, a bit of advice to watch the conversions between `[]byte` and `string` was offered, that it was often unnecessary and done only because we are accustomed to working with strings, and that in high-throughput scenarios, it exacts a performance penalty. A couple years later, this seems to have been relevant advice. For example, the most effective way to work with things in Go is with `io.Reader` and `io.Writer`, interfaces to which byte slices lend themselves pretty nicely. Perhaps it is necessary in this case to have a string, for a reason that is not clear to me without additional context, but it is worth planting the seed to be mindful of conversions to `string`.